### ISSUE TITLE
Correct way to initialize a model on TPUs

### ISSUE THREAD

**Issue by igrekun** [google/flax#1277]
When module init call is wrapped with jax.jit XLA compiler takes a long time to create the model.

Medium sized transformers with 512 dim and 12 layers hit compilation time limit of 1 hour.
Without JIT the model compiles in seconds, but examples do wrap the call.

What is the correct way to initialize the model and what's the reasoning behind wrapping the call with jax.jit?

EDIT: I expanded the example to show that jitted model results in NaN outputs for larger models, while non-jit model works fine. Could that be a compiler bug?

Reproducible example modified from WMT Flax template:
https://colab.research.google.com/drive/1wIE4h4duj0J5dX0angN2fm0vFkRLPqa2?usp=sharing

**Comment by igrekun** [comment]
Using WMT example the issue with difference between jit and non jit versions can be illustrated by these lines:

```
key = random.PRNGKey(42)
tr_config = TransformerConfig(vocab_size=128, output_vocab_size=128, num_layers=24, deterministic=True, posemb_init=nn.initializers.normal(stddev=0.02))
initial_variables = Encoder(tr_config).init(key, jnp.ones((16, 128), dtype=jnp.int32))

@jax.jit
def jit_apply(vars, x):
    return Encoder(tr_config).apply({"params": vars["params"]}, x)

def simple_apply(vars, x):
    return Encoder(tr_config).apply({"params": vars["params"]}, x)

jit_output = jit_apply(initial_variables, jnp.ones((16, 128)))
simple_output = simple_apply(initial_variables, jnp.ones((16, 128)))

print(simple_output[0,0,:10])
print(jit_output[0,0,:10])
```

which outputs

[-0.02371583  1.0241227   1.1573081  -0.69314075  0.00546649  0.47282943
 -0.1721289   2.0183356  -1.8775052   1.3018961 ]
[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]

**Comment by igrekun** [comment]
I found the culprit. Somehow passing params dict to jitted function breaks the execution. 

Snippet below works fine:

```
key = random.PRNGKey(42)
tr_config = TransformerConfig(vocab_size=128, output_vocab_size=128, num_layers=12, deterministic=True, posemb_init=nn.initializers.normal(stddev=0.02))
initial_variables = Encoder(tr_config).init(key, jnp.ones((16, 128), dtype=jnp.int32))

@jax.jit
def ext_apply(x):
    return Encoder(tr_config).apply({"params": initial_variables["params"]}, x)

bound_output = ext_apply(jnp.ones((16, 128)))
```

While that solves the problem, it is still interesting to know why this happens and why one should or should not JIT the init call.

**Comment by jheek** [comment]
This looks like a compilation bug in XLA. What device did you use?

**Comment by avital** [comment]
Hi @igrekun -- this may be a JAX or XLA bug. Behavior under jit should not change the results in this way. Could you share a simple stand-alone repro?

Also your second example (where you close over `init_variables`) isn't good -- that way the variables are compiled as constants leading to really larged compiled programs, as well as inability for the jitted function to behave different when parameters change.

**Comment by igrekun** [comment]
Used the TPU. 

Added the simple example at the end of the original colab that I linked. Here it is
https://colab.research.google.com/drive/1wIE4h4duj0J5dX0angN2fm0vFkRLPqa2?usp=sharing

The weird part is that the bug manifests itself when the model becomes large enough. Under 6 layer depth it works fine.

If colab is not convinient here is all the code

```
class SimpleModel(nn.Module):
    depth: int = 12
    @nn.compact
    def __call__(self, x):
        for i in range(self.depth):
            residual = x
            x = nn.LayerNorm()(x)
            x = nn.SelfAttention(
                num_heads=8,
                dtype=jnp.float32,
                qkv_features=512,
                deterministic=True,
            )(x)
            x = x + residual
        return x

def get_model(input_shape):
    key = random.PRNGKey(42)
    return SimpleModel().init(key, jnp.ones(input_shape, dtype=jnp.float32))["params"]

@jax.jit
def test_jit_preds(params, x):
    return SimpleModel().apply({"params": params}, x)

def test_plain_preds(params, x):
    return SimpleModel().apply({"params": params}, x)

output = test_jit_preds(params, jnp.ones((128, 128, 512)))
print(output[0,0,:10])
output = test_plain_preds(params, jnp.ones((128, 128, 512)))
print(output[0,0,:10])
```

The output is
[nan nan nan nan nan nan nan nan nan nan]
[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]

**Comment by jheek** [comment]
Could you try initialising with random noise like `random.normal(random.PRNGKey(1), (128, 128, 512))`?

**Comment by igrekun** [comment]
Expanded example to include noise init as well as noise test input. Doesn't seem to make NaNs go away though ):

**Comment by jheek** [comment]
@igrekun are you hunting down XLA bugs or are just having bad luck today? ;)

I cannot reproduce this issue on a normal TPU machine also so this also seems related to Colab TPU somehow.

**Comment by igrekun** [comment]
@jheek XLA proved to compile more efficient code for the type of models I'm working with so I went to porting all my code from pytorch to JAX.

I guess just bad luck (:

Also, seems like this one might be related to the recent one on JAX repo
https://github.com/google/jax/issues/6558

Seems very similar to me.

**Comment by jheek** [comment]
Ah good catch it's probably that. Have you tried the single device pmap workaround? In that case we would be certain

**Comment by igrekun** [comment]
Yeah, it is the case. Single device pmap produces matching outputs. The code is at the end of the original colab notebook. (:

**Comment by jheek** [comment]
Nice, I'll close this issue because it will be fixed int the upcoming jaxlib release
